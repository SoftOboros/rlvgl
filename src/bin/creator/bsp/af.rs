//! Alternate-function lookup trait used by the BSP tests.
//!
//! Implementations provide a mapping from (MCU, pin, signal) to
//! an alternate function number. The tests supply a stub provider
//! with hard-coded values.

use std::collections::HashMap;

/// Trait returning an alternate function number for a pin/signal pair.
pub trait AfProvider {
    /// Lookup the alternate function number for `pin` serving `func` on `mcu`.
    /// Returns `None` if the function is not available on that pin.
    fn lookup_af(&self, mcu: &str, pin: &str, func: &str) -> Option<u8>;
}

/// JSON-backed alternate-function database used by integration tests.
///
/// The JSON structure matches the output of
/// `tools/afdb/st_extract_af.py`:
/// `{ "MCU": { "PIN": { "SIGNAL": AF }}}`.
#[allow(dead_code)]
pub struct JsonAfDb {
    map: HashMap<String, HashMap<String, HashMap<String, u8>>>,
}

impl JsonAfDb {
    /// Load the database from `path`.
    #[allow(dead_code)]
    pub fn from_path(path: &std::path::Path) -> anyhow::Result<Self> {
        let data = std::fs::read_to_string(path)?;
        let map = serde_json::from_str(&data)?;
        Ok(Self { map })
    }
}

impl AfProvider for JsonAfDb {
    fn lookup_af(&self, mcu: &str, pin: &str, func: &str) -> Option<u8> {
        self.map
            .get(mcu)
            .and_then(|p| p.get(pin))
            .and_then(|m| m.get(func))
            .copied()
    }
}
