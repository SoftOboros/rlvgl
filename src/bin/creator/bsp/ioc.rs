//! Minimal STM32 CubeMX `.ioc` parser used in BSP tests.
//!
//! The parser extracts pin signals, PLL parameters and kernel clock
//! selections without relying on vendor-maintained tables. Alternate
//! function numbers are supplied by an [`AfProvider`].

use crate::af::AfProvider;
use crate::ir::{Clocks, Ir, Peripheral, Pin, Pll};
use anyhow::{Result, anyhow};
use indexmap::IndexMap;
use indexmap::IndexMap as HashMap;
use regex::Regex;

/// Convert raw `.ioc` text into the [`Ir`] representation.
///
/// Errors if a reserved pin (PA13/PA14) is configured and `allow_reserved`
/// is false.
pub fn ioc_to_ir(text: &str, af: &dyn AfProvider, allow_reserved: bool) -> Result<Ir> {
    let kv = parse_kv(text);
    let mcu = kv.get("Mcu.Name").cloned().unwrap_or_default();
    let package = kv.get("Mcu.Package").cloned().unwrap_or_default();

    let pin_re = Regex::new(r"^P([A-Z])(\d+)\.Signal$").unwrap();
    let kernel_re = Regex::new(r"^RCC\.([A-Za-z0-9]+)ClockSelection$").unwrap();
    let mut pins = Vec::new();
    let mut peripherals: IndexMap<String, Peripheral> = IndexMap::new();
    let mut kernels: IndexMap<String, String> = IndexMap::new();

    for (k, v) in kv.iter() {
        if let Some(caps) = pin_re.captures(k) {
            let pin = format!("P{}{}", &caps[1], &caps[2]);
            let func = v.to_string();
            let afn = af.lookup_af(&mcu, &pin, &func).unwrap_or(0);
            pins.push(Pin {
                pin: pin.clone(),
                func: func.clone(),
                af: afn,
            });
            if let Some((name, class, role)) = split_signal(&func) {
                let periph = peripherals.entry(name).or_insert(Peripheral {
                    class,
                    signals: IndexMap::new(),
                });
                periph.signals.insert(role, pin);
            }
        } else if let Some(caps) = kernel_re.captures(k) {
            let periph = caps[1].to_lowercase();
            if let Some(src) = v.rsplit('_').next() {
                kernels.insert(periph, src.to_lowercase());
            }
        }
    }

    let mut pll_map: IndexMap<String, Pll> = IndexMap::new();
    for idx in 1..=3 {
        let m = kv.get(&format!("RCC.PLL{}M", idx));
        let n = kv.get(&format!("RCC.PLL{}N", idx));
        let p = kv.get(&format!("RCC.PLL{}P", idx));
        let q = kv.get(&format!("RCC.PLL{}Q", idx));
        let r = kv.get(&format!("RCC.PLL{}R", idx));
        if let (Some(m), Some(n), Some(p), Some(q), Some(r)) = (m, n, p, q, r) {
            pll_map.insert(
                format!("pll{}", idx),
                Pll {
                    m: m.parse()?,
                    n: n.parse()?,
                    p: p.parse()?,
                    q: q.parse()?,
                    r: r.parse()?,
                },
            );
        }
    }

    let clocks = Clocks {
        pll: pll_map,
        kernels,
    };

    if !allow_reserved {
        for p in &pins {
            if p.pin == "PA13" || p.pin == "PA14" {
                return Err(anyhow!("reserved pin {} configured", p.pin));
            }
        }
    }

    Ok(Ir {
        mcu,
        package,
        clocks,
        pinctrl: pins,
        peripherals,
    })
}

fn parse_kv(s: &str) -> HashMap<String, String> {
    s.lines()
        .filter_map(|l| l.split_once('='))
        .map(|(k, v)| (k.trim().to_string(), v.trim().to_string()))
        .collect()
}

fn split_signal(sig: &str) -> Option<(String, String, String)> {
    // Allow class prefixes with embedded digits (e.g., I2C1_SCL).
    let re = Regex::new(r"^([A-Z0-9]+?)(\d+)_([A-Z0-9]+)$").unwrap();
    if let Some(caps) = re.captures(sig) {
        let class_raw = &caps[1];
        let inst = &caps[2];
        let role_raw = &caps[3];
        let class = match class_raw {
            "USART" | "UART" => "serial",
            "SPI" => "spi",
            "I2C" => "i2c",
            _ => return None,
        };
        Some((
            format!("{}{}", class_raw.to_lowercase(), inst),
            class.to_string(),
            role_raw.to_lowercase(),
        ))
    } else {
        None
    }
}
