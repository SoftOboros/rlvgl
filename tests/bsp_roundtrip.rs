//! Tests the BSP generation pipeline: `.ioc` â†’ IR â†’ template.
#![cfg(all(feature = "creator", feature = "regression"))]

#[path = "../src/bin/creator/bsp/af.rs"]
mod af;
#[path = "../src/bin/creator/bsp/ioc.rs"]
mod ioc;
#[path = "../src/bin/creator/bsp/ir.rs"]
mod ir;

use af::JsonAfDb;
use ioc::ioc_to_ir;
use minijinja::{Environment, context};
use std::{path::PathBuf, process::Command};
use tempfile::NamedTempFile;

#[test]
fn ioc_roundtrip_snapshot() {
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let csv = manifest_dir.join("tests/fixtures/stm32_af.csv");
    let script = manifest_dir.join("tools/afdb/st_extract_af.py");
    let tmp = NamedTempFile::new().unwrap();
    let status = Command::new("python3")
        .arg(&script)
        .arg("--db")
        .arg(&csv)
        .arg("--out")
        .arg(tmp.path())
        .status()
        .expect("failed to run st_extract_af.py");
    assert!(status.success());
    let afdb = JsonAfDb::from_path(tmp.path()).unwrap();

    let ioc_text = include_str!("fixtures/simple.ioc");
    let ir = ioc_to_ir(ioc_text, &afdb, false).unwrap();
    insta::assert_yaml_snapshot!("ir", &ir);

    let mut env = Environment::new();
    env.add_template(
        "gen",
        include_str!("../src/bin/creator/bsp/templates/simple.rs.jinja"),
    )
    .unwrap();
    let tmpl = env.get_template("gen").unwrap();
    let rendered = tmpl.render(context! { spec => &ir }).unwrap();
    insta::assert_snapshot!("generated", rendered);
}
