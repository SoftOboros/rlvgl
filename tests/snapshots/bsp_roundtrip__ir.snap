---
source: tests/bsp_roundtrip.rs
expression: "&ir"
---
mcu: STM32H747XIHx
package: LQFP176
clocks:
  pll:
    pll1:
      m: 5
      n: 400
      p: 2
      q: 4
      r: 2
  kernels:
    usart1: pclk2
pinctrl:
  - pin: PA9
    func: USART1_TX
    af: 7
  - pin: PA10
    func: USART1_RX
    af: 7
  - pin: PB0
    func: GPIO_Output
    af: 0
  - pin: PB3
    func: SPI1_SCK
    af: 5
  - pin: PB4
    func: SPI1_MISO
    af: 5
  - pin: PB5
    func: SPI1_MOSI
    af: 5
  - pin: PB6
    func: I2C1_SCL
    af: 4
  - pin: PB7
    func: I2C1_SDA
    af: 4
peripherals:
  usart1:
    class: serial
    signals:
      tx: PA9
      rx: PA10
  spi1:
    class: spi
    signals:
      sck: PB3
      miso: PB4
      mosi: PB5
  i2c1:
    class: i2c
    signals:
      scl: PB6
      sda: PB7
